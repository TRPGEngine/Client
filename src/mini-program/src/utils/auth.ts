import { request } from '@shared/utils/request';
import Taro from '@tarojs/taro';
import _last from 'lodash/last';

/**
 * 发送一个校验token的请求到服务端
 * 如果校验不通过会由request的拦截器自动跳转到登录页
 * 该检查一般用于表单填写页面防止填了一堆数据后提交失败跳转到登录页的挫败感
 * @param 是否自动跳转。如果为登录页则此值应该为false
 */
export async function checkTaroToken(autoNav = true): Promise<boolean> {
  try {
    await request.post('/player/sso/check');
    return true;
  } catch (err) {
    if (autoNav) {
      // 如果出错则跳转到登录页面
      const currentPages = Taro.getCurrentPages();
      if (currentPages.length === 1) {
        Taro.navigateTo({
          url: `/pages/login/index`,
        });
      } else {
        const lastPage = _last(currentPages);
        Taro.redirectTo({
          url: `/pages/login/index?next=${lastPage?.route ?? ''}`,
        });
      }

      return false;
    }
    throw err;
  }
}
