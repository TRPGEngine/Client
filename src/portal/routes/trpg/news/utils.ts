import _isString from 'lodash/isString';
import _get from 'lodash/get';

const imageUrlRE = /<img.*?src="(.*?)"/;

/**
 * 从HTML中提取第一张图片
 * @param html HTML字符串
 */
export function getFirstImageUrlFromHTML(html: string): string | null {
  const res = html.match(imageUrlRE);
  const url = _get(res, [1]);

  if (_isString(url)) {
    return url;
  }

  return null;
}
