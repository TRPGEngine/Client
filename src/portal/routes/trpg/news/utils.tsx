import { Parser } from 'html-to-react';

/**
 * Document:
 * https://www.npmjs.com/package/html-to-react
 */
const parser = new Parser();

/**
 * 将 html 转换为适合展示的文本
 */
export function transformToDom(html: string): React.ReactNode {
  return parser.parse(html);
}
