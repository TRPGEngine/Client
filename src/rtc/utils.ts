import type { RTCStoreType } from './redux';
import _mapValues from 'lodash/mapValues';
import _values from 'lodash/values';
import { Unsubscribe } from '@reduxjs/toolkit';

let isWatching = false;
let unsubscribeWatchingFn: Unsubscribe = () => {};

function buildEventFn(
  store: RTCStoreType,
  onAddTrack: (newTrack: MediaStreamTrack) => void
) {
  const alreadyTrackList: MediaStreamTrack[] = []; // 这是已经存储在列表中的音频轨道
  return () => {
    if (isWatching === false) {
      return;
    }

    const allAudioTracks = _values(store.getState().consumers)
      .map((consumer) => {
        return consumer.track;
      })
      .filter((track) => track.kind === 'audio');

    for (const track of allAudioTracks) {
      if (!alreadyTrackList.includes(track)) {
        alreadyTrackList.push(track);
        onAddTrack(track);
      }
    }
  };
}

export function startWatchAllAudioTrack(
  store: RTCStoreType,
  onAddTrack: (newTrack: MediaStreamTrack) => void
) {
  if (isWatching === true) {
    console.log('[startWatchAllAudioTrack] skip because is watching');
  }

  isWatching = true;
  const fn = buildEventFn(store, onAddTrack);
  fn(); // 开始时先调用一下
  unsubscribeWatchingFn = store.subscribe(fn);
}

export function stopWatchAllAudioTrack() {
  isWatching = false;
  unsubscribeWatchingFn();
}
