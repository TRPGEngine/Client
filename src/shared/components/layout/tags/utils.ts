import { Fragment } from 'react';
import _get from 'lodash/get';
import _set from 'lodash/set';
import _isNil from 'lodash/isNil';
import _clone from 'lodash/clone';
import _toPairs from 'lodash/toPairs';
import _fromPairs from 'lodash/fromPairs';
import type {
  XMLBuilderState,
  XMLBuilderContext,
} from '@shared/components/layout/XMLBuilder';
import {
  StateDataType,
  StateActionType,
} from '@shared/components/layout/types';
import type { XMLElement } from '../parser/xml-parser';

export type OperationDataType = {
  scope: string;
  field: string;
};

/**
 * 获取需要操作的变量的作用域与操作的变量名
 * 作用域默认为data
 * 用于独立一个path的第一段
 * example:
 * - 'a' => {scope: 'data', field: 'a'}
 * - 'a.b' => {scope: 'a', field: 'b'}
 * - 'a.b.c' => {scope: 'a', field: 'b.c'}
 *
 * @param str 操作参数字符串
 */
export const getOperationData = (str: string): OperationDataType => {
  const [scope, ...fields] = str.split('.');

  if (fields.length > 0) {
    // 如果为abc.def
    return {
      scope,
      field: fields.join('.'),
    };
  } else {
    // 如果为abc
    return {
      scope: 'data',
      field: scope,
    };
  }
};

/**
 * 尝试将文本转化为数字
 * 用于表单数据转换
 * @param str 要转换的文本
 */
export const tryToNumber = (str: string): number | string => {
  if (typeof str === 'string' && str === '') {
    // 空字符串直接返回空字符串
    return '';
  }

  if (typeof str === 'string' && (str.length > 15 || str.startsWith(' '))) {
    // 大于15位的数字视为字符串
    // 输入空格视为字符串
    return str;
  }

  const num = Number(str);
  return !isNaN(num) ? num : str;
};

/**
 * 标准化标签名。返回字符串或空节点
 * @param tagName 标签名
 */
export const normalizeTagName = (tagName: string): string | typeof Fragment => {
  // 如果是首字母开头。视为没有做定义的内置操作。改为t-xxx的格式防止抛出命名警告
  if (typeof tagName === 'string' && /[A-Z]/.test(tagName[0])) {
    tagName = 't-' + tagName.toLowerCase();
  }

  // 如果是空字符串或者undefined。使用react的Fragment
  if (!tagName) {
    return Fragment;
  }

  return tagName;
};

/**
 * 获取要实际操作变量名的路径
 *
 * 返回要获取变量在data数据中的位置
 */
export const getNamePath = (name: string) => {
  const { scope, field } = getOperationData(name);
  let path = field;
  if (scope !== 'data') {
    path = [scope, field].join('.');
  }

  return path;
};

/**
 * 根据上下文获取指定状态的数据
 */
export const getStateValue = (
  context: XMLBuilderContext,
  bindingName: string
): StateDataType => {
  const path = getNamePath(bindingName);

  let ret = _get(context.state.data, path);
  if (_isNil(ret)) {
    // 如果没有从data中取到。可能path是有前缀的路径。尝试从上一级获取
    ret = _get(context.state, path);
  }

  // NOTICE: 不知道为什么要不允许获取object
  // 直接返回匹配的数据，如果有问题再改回来
  // if (typeof ret === 'object') {
  //   return '';
  // }

  return ret as any;
};

/**
 * 设置状态的值
 */
export const setStateValue = (
  context: XMLBuilderContext,
  bindingName: string,
  value: StateDataType
): void => {
  const { dispatch } = context;
  const { scope, field } = getOperationData(bindingName);

  dispatch({
    type: StateActionType.UpdateData,
    payload: {
      scope,
      field,
      value,
    },
  });
};
/**
 * 将属性中额私有参数提取出来(以下划线开头_的成员变量)
 * 返回一个新的对象
 */
export const removePrivateProps = <T extends object>(props: T): Partial<T> => {
  return _fromPairs(
    _toPairs(props).filter(([key, value]) => !key.startsWith('_'))
  ) as T;
};

/**
 * 处理xml读取到的多行文本
 */
export const parseMultilineText = (text: string) => {
  // 支持\n的渲染 拿到的换行符为\\n
  if (_isNil(text) || typeof text !== 'string') {
    return '';
  }

  return text.replace(new RegExp('\\\\n', 'g'), '\n');
};

/**
 * 获取元素子节点的文本
 */
export const getChildrenText = (el: XMLElement): string => {
  const childrenEl = el.elements!;

  return childrenEl
    .map<string>((el) => {
      if (el.type === 'text') {
        return String(el.text);
      } else if (el.type === 'element') {
        return getChildrenText(el);
      } else {
        return '';
      }
    })
    .reduce((prev, curr) => {
      return prev + curr;
    }, '');
};
