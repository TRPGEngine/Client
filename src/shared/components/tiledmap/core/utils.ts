import type { Size, Position, Rect } from './types';

/**
 * 计算距离数最近的网格点
 * @param num 真实位置
 * @param gridSizeNum 网格大小
 */
function calcPx2GridMagnetNum(num: number, gridSizeNum: number): number {
  return Math.round(num / gridSizeNum);
}

/**
 * 计算距离数最左上角的点
 * @param num
 * @param gridSizeNum
 */
function calcPx2GridFloorNum(num: number, gridSizeNum: number): number {
  return Math.floor(num / gridSizeNum);
}

/**
 * 像素位置转网格位置
 */
export function px2gridPos(
  pos: Position,
  gridSize: Size,
  isMagnet: boolean = false
): Position {
  const calcFn = isMagnet ? calcPx2GridMagnetNum : calcPx2GridFloorNum;

  return {
    x: calcFn(pos.x, gridSize.width),
    y: calcFn(pos.y, gridSize.height),
  };
}

/**
 * 标准化Rect
 */
export function normalizeRect(rect: Rect): Rect {
  return {
    x1: Math.min(rect.x1, rect.x2),
    y1: Math.min(rect.y1, rect.y2),
    x2: Math.max(rect.x1, rect.x2),
    y2: Math.max(rect.y1, rect.y2),
  };
}

/**
 * 判断rect1是否在rect2中
 */
export function isRectInsideRect(rect1: Rect, rect2: Rect) {
  rect1 = normalizeRect(rect1);
  rect2 = normalizeRect(rect2);

  return (
    rect1.x1 >= rect2.x1 &&
    rect1.x2 <= rect2.x2 &&
    rect1.y1 >= rect2.y1 &&
    rect1.y2 <= rect2.y2
  );
}

/**
 * 将pxRect转化成gridRect
 * 算法是网里缩
 */
export function pxRect2GridRect(rect: Rect, gridSize: Size): Rect {
  rect = normalizeRect(rect);

  return {
    x1: Math.ceil(rect.x1 / gridSize.width),
    y1: Math.ceil(rect.y1 / gridSize.height),
    x2: Math.floor(rect.x2 / gridSize.width),
    y2: Math.floor(rect.y2 / gridSize.height),
  };
}

/**
 * 判断一个点是否在一个矩形空间内
 */
export function isPositionInsizeRect(pos: Position, rect: Rect): boolean {
  rect = normalizeRect(rect);

  return (
    pos.x <= rect.x2 && pos.x >= rect.x1 && pos.y <= rect.y2 && pos.y >= rect.y1
  );
}

/**
 * 根据宽高和中心点返回一个矩形
 */
export function getRectWithSize(pos: Position, size: Size): Rect {
  return {
    x1: pos.x - size.width / 2,
    y1: pos.y - size.height / 2,
    x2: pos.x + size.width / 2,
    y2: pos.y + size.height / 2,
  };
}
