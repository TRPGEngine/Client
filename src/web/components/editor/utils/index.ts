import { Editor, Transforms, Range, Element } from 'slate';
import type { FormatType, TRPGEditor, TRPGEditorValue } from '../types';

const LIST_TYPES = ['numbered-list', 'bulleted-list'];

/**
 * 创建一个空的slate数据对象
 */
export function buildBlankInputData(): TRPGEditorValue[] {
  return [
    {
      children: [{ text: '' }],
    } as TRPGEditorValue,
  ];
}

/**
 * 获取文档开始位置
 */
export function getHeadSelection(): Range {
  return {
    anchor: { path: [0, 0], offset: 0 },
    focus: { path: [0, 0], offset: 0 },
  };
}

/**
 * 判定块级元素是否生效
 */
export const isBlockActive = (editor: TRPGEditor, format: FormatType) => {
  const [match] = Array.from(
    Editor.nodes(editor, {
      match: (n) =>
        !Editor.isEditor(n) && Element.isElement(n) && n.type === format,
    })
  );

  return !!match;
};

/**
 * 切换块级元素属性
 */
export const toggleBlock = (editor: TRPGEditor, format: FormatType) => {
  const isActive = isBlockActive(editor, format);
  const isList = LIST_TYPES.includes(format);

  Transforms.unwrapNodes(editor, {
    match: (n) =>
      !Editor.isEditor(n) &&
      Element.isElement(n) &&
      LIST_TYPES.includes(n.type),
    split: true,
  });

  Transforms.setNodes(editor, {
    type: isActive ? 'paragraph' : isList ? 'list-item' : format,
  });

  if (!isActive && isList) {
    const block = { type: format, children: [] };
    Transforms.wrapNodes(editor, block);
  }
};

/**
 * 判断标记是否生效
 */
export const isMarkActive = (editor: TRPGEditor, format: string) => {
  const marks = Editor.marks(editor);
  return marks ? marks[format] === true : false;
};

/**
 * 切换标记
 */
export const toggleMark = (editor: TRPGEditor, format: string) => {
  const isActive = isMarkActive(editor, format);

  if (isActive) {
    Editor.removeMark(editor, format);
  } else {
    Editor.addMark(editor, format, true);
  }
};
