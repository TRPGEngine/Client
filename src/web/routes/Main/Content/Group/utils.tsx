import type { GroupPanelType } from '@shared/types/panel';
import { Iconfont } from '@web/components/Iconfont';
import React from 'react';
import styled from 'styled-components';

const DefaultPanelIconContainer = styled.div`
  padding-left: 2px;
  padding-right: 2px;
`;

/**
 * 获取团面板Icon
 */
export function getGroupPanelIcon(panelType: GroupPanelType) {
  if (panelType === 'voicechannel') {
    return <Iconfont>&#xe664;</Iconfont>;
  } else if (panelType === 'note') {
    return <Iconfont>&#xe621;</Iconfont>;
  } else if (panelType === 'website') {
    return <Iconfont>&#xe651;</Iconfont>;
  } else if (panelType === 'calendar') {
    return <Iconfont>&#xe867;</Iconfont>;
  }

  return <DefaultPanelIconContainer>#</DefaultPanelIconContainer>;
}
