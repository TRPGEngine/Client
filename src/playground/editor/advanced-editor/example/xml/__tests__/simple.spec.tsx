import { render } from '@testing-library/react';
import React from 'react';
import fs from 'fs';
import path from 'path';
import { sleep } from '@test/utils';
import { XMLBuilderTester } from '@shared/components/layout/__tests__/utils';

const layout = String(
  fs.readFileSync(path.resolve(__dirname, '../simple.xml'))
);

test('basic snapshot', async () => {
  const { container } = render(<XMLBuilderTester xml={layout} />);

  await sleep(500);

  expect(container).toMatchSnapshot();
});
