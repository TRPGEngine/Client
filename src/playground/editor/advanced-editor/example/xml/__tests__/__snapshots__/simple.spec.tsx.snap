// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic snapshot 1`] = `
.c3 {
  margin-top: 0.5rem;
  margin-bottom: 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-right: 4px;
}

.c5 {
  display: inline-block;
  max-width: 100%;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c1 {
  padding: 10px;
}

.c2 {
  -webkit-flex-wrap: wrap-reverse !important;
  -ms-flex-wrap: wrap-reverse !important;
  flex-wrap: wrap-reverse !important;
  margin-bottom: 10px;
  border-bottom: 1px solid rgba(232,232,232,0.8);
  padding-top: 24px;
}

.c0 {
  text-align: left;
  width: 100%;
}

.c0 pre {
  margin: 0;
}

.c0 code {
  padding: 0.2em 0.4em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(27,31,35,0.05);
  border-radius: 3px;
}

<div>
  <div
    class="c0"
    style="position: relative;"
  >
    <iframe
      style="background: transparent; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%; z-index: -1;"
    />
    <div
      class="c1"
    >
      <div
        class="ant-row c2"
        type="flex"
      >
        <div
          class="ant-col ant-col-24"
        >
          <form
            class="ant-form ant-form-vertical "
          >
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-col-6 ant-form-item-label"
              >
                <label
                  class="ant-form-item-required"
                  title="名称"
                >
                  名称
                </label>
              </div>
              <div
                class="ant-col ant-col-18 ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <mock-antd-input />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="ant-row ant-form-item"
            >
              <div
                class="ant-col ant-col-6 ant-form-item-label"
              >
                <label
                  class=""
                  title="描述"
                >
                  描述
                </label>
              </div>
              <div
                class="ant-col ant-col-18 ant-form-item-control"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <mock-antd-input-textarea
                      autosize="[object Object]"
                    />
                  </div>
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          class="ant-col ant-col-24"
          style="text-align: center; margin-bottom: 10px;"
        >
          <div>
            <div
              style="cursor: pointer; display: inline-block; position: relative;"
            >
              <input
                accept="image/*"
                style="display: none;"
                type="file"
              />
              <span
                class="ant-avatar ant-avatar-circle ant-avatar-image ant-avatar-icon"
                style="width: 64px; height: 64px; line-height: 64px; font-size: 25.6px; cursor: inherit; user-select: none;"
              >
                <img
                  src="undefined"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        _childrenel="[object Object]"
        _el="[object Object]"
        _name="BaseAttr"
        class="ant-col ant-col-24"
      >
        <div
          class="ant-row c3 tag-row"
        >
          <div
            class="ant-col ant-col-6"
          >
            <div
              class="c4"
            >
              <pre
                class="c5"
              >
                人物卡信息
              </pre>
               
            </div>
          </div>
          <div
            class="ant-col ant-col-18"
          >
            <mock-antd-input-textarea
              autosize="true"
              placeholder="人物卡信息"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
